README.md
# pocketool
The official github page for pocketool! 
_____________________________________________________________________________________
[<img src="https://discordapp.com/api/guilds/486910899756728320/widget.png?style=shield">](https://discord.gg/d3zRBsc)
_____________________________________________________________________________________
*Pocketool* is a discord.py rewrite bot, made for utility. It's (planned) features are (projected to have) serverinfo, userinfo, roleinfo, basic moderation, server creation and management commands and templates, and raid protection.
_____________________________________________________________________________________
Currently, the bot is in beta.  with basic commands, im not expecting it to get servers fast.
**At the time of this commit, the bot is in _closed_ beta, meaning only i can add it to servers should they apply for beta.**
_____________________________________________________________________________________
I, 'EEK', do not claim responsibility for any damage(s) caused by the bot - the code (and usage of the bot) are used at your own risk.
_____________________________________________________________________________________
"WhErEs ThE CoDe?!?!/!?!?1/1/1/?"
*with the bot being in beta, i haven't posted the code. it will be released at open beta, or if the [server's poll](https://discord.gg/gmvNT4Y) says to make it OS now.*
_____________________________________________________________________________________


changelog.md
# 14/3/19
didnt do much, added git pages and broke a lot of stuff.
kinda fixed userinfo
# some date
i tried to fix webhooks.


pock/boot.py
import discord, time, datetime, os, traceback
from discord.ext import commands
start = time.time()
bot = commands.AutoShardedBot(command_prefix=commands.when_mentioned_or('//'), description=None, max_messages=9999,
                              case_insensitive=True, activity=discord.Game(name="Booting..."),
                              status=discord.Status.do_not_disturb)

token = 'PUT_YOUR_TOKEN_HERE'
toload = ["coggable.listeners",
                       "coggable.maincog",
                       "coggable.help",
                       "coggable.role"]
bot.cogss = toload
bot.boot_time = 0
bot.loadedcogs = 0
@bot.event
async def on_ready():
    success = 0
    for c in toload:
        try:
            bot.load_extension(c)
            bot.loadedcogs += 1
            success += 1
        except:
            pass
    done = time.time()
    bt = round(done - start)
    bot.boot_time += bt
    print("__________________________________________")
    print("Logged in as: {}".format(bot.user.name))
    print("Total guilds: {}".format(len(bot.guilds)))
    print("Total users : {}".format(len(bot.users)))
    print("Boot time   : {}s".format(bt))
    print("Total loaded: {}".format(success))
    print("__________________________________________")
    await bot.change_presence(activity=discord.Game(name='with {} servers'.format(len(bot.guilds))))

@bot.command()
async def test(ctx):
    """| test smthn"""
    await ctx.send("Hello {}! todays date is {}, and the UTC time is {}! i have received `{}` "
                   "messages since "
                   "boot.".format(ctx.author.mention, datetime.datetime.utcnow().strftime('%d/%m/%Y'),
                                  datetime.datetime.utcnow().strftime('%I:%M %p'), len(ctx.bot.messages)))


@bot.command()
@commands.is_owner()
async def reboot(ctx):
    """ | Reboot me"""
    try:
        await ctx.message.add_reaction("success:522078924432343040")
        await bot.close()
    except:
        pass
    finally:
        os.system("python3 boot.py")


@bot.command(aliases=['reload'])
@commands.is_owner()
async def r(ctx, cog: str):
    """| re/load a cog"""
    try:
        if cog == 'all':
            successful = ""
            for c in toload:
                try:
                    bot.unload_extension(c)
                    bot.load_extension(c)
                    successful += "Reloaded {} <:success:522078924432343040>\n".format(c)
                except:
                    print(f'{traceback.format_exc()}__________________________________________')
                    successful += "couldn't reload {} <:fail:522076877075251201>\n".format(c)
                    continue
            await ctx.send(successful)
        else:
            bot.unload_extension(cog)
            bot.load_extension(cog)
            await ctx.send("Reloaded {} <:success:522078924432343040>\n".format(cog), delete_after=10)
        await ctx.message.add_reaction("success:522078924432343040")
    except Exception as e:
        await ctx.message.add_reaction('fail:522076877075251201')
        print(f'{traceback.format_exc()}__________________________________________')
        await ctx.send(f'```py\n{e}\n```Full error in console')


@bot.command(aliases = ['ping', 'uptime'])
async def shard(ctx):
    """Get your shard, ping and uptime."""
    current_time = time.time()
    difference = int(round(current_time - start, 2))

    seconds = difference % 60
    minutes = round(difference / 60)
    hours = round(difference / 3600)
    days = 0

    while minutes > 59:
        minutes -= 60
        hours += 1
    while hours >= 23:
        days += 1
        hours -= 24
    utfmt = f"{days}d, {hours}h, {minutes}m, {seconds}s"
    m = await ctx.send("Getting info...")
    async with ctx.channel.typing():
        pang = f'{ctx.bot.latency * 1000:.0f}'
        shardid = ctx.guild.shard_id + 1
        e = discord.Embed(title="Total shards: {}".format(bot.shard_count),
                          description="**Your shard: {}**\n**Ping: {}ms**"
                                      "\n**Boot time: {}s**\nUptime: **{}**".format(shardid, pang, bot.boot_time, utfmt),
                          color=ctx.author.color, timestamp=datetime.datetime.utcnow())
    await ctx.send(embed=e)
    await m.delete()

bot.run(token)


pock/coggable/help.py
# -*- coding: utf-8 -*-

"""
The MIT License (MIT)
Copyright (c) 2018-2019 laggycomputer
Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
"""

import discord
from discord.ext import commands

class Help(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.command(description="Please note that if you specify a cog name that is also the name of a command, "
                                  "the help for the command will be given.")
    async def help(self, ctx, specific=None):
        """Shows this message."""
        speific = specific
        if specific:

            command = self.bot.get_cog(speific)
            command = self.bot.get_command(specific)

            if command is None:
                await ctx.send(":x: Command/cog not found!")
                return
            else:
                if isinstance(command, commands.Cog):
                    cog = command
                    del command

                    emb = discord.Embed(title=f"{ctx.prefix}{cog.name}", color=ctx.author.color)

                    for command in cog.walk_commands():
                        alist = []
                        for alias in command.aliases:
                           alist.append(f"`{ctx.prefix}{alias}`")

                        emb.add_field(name=f"`{ctx.prefix}{command.qualified_name}`{', ' if any(command.aliases) else ''}{', '.join(alist)}", description=command.short_doc + "\u200b", inline=False)
                else:
                    alist = []
                    for alias in command.aliases:
                        alist.append(f"`{alias}`")

                    emb = discord.Embed(title=f"{ctx.prefix}{command.qualified_name}{', ' if any(command.aliases) else ''}{', '.join(alist)}", description=command.short_doc, color=ctx.author.color)
                    emb.add_field(name="Full description", value=command.description + "\u200b")

                    if isinstance(command, commands.GroupMixin):
                        for subcommand in command.commands:
                            alist = []
                            for alias in subcommand.aliases:
                                alist.append(f"`{ctx.prefix}{alias}`")

                            emb.add_field(name=f"`{ctx.prefix}{subcommand.qualified_name}`{', ' if any(subcommand.aliases) else ''}{', '.join(alist)}", value=subcommand.short_doc + "\u200b", inline=False)
        else:
            emb = discord.Embed(title="Help and Information", color=0xf92f2f)
            for name, command in self.bot.all_commands.items():
                if name in command.aliases or (command.hidden and not await self.bot.is_owner(ctx.author)):
                    continue

                alist = []
                for alias in command.aliases:
                    alist.append(f"`{ctx.prefix}{alias}`")

                emb.add_field(name=f"{name}{', ' if any(command.aliases) else ''}{', '.join(alist)}", value=command.short_doc + "\u200b", inline=False)

        emb.set_thumbnail(url=self.bot.user.avatar_url)
        emb.set_author(name=self.bot.user.name, icon_url=self.bot.user.avatar_url, url='https://invite.gg/ebot')

        await ctx.send(embed=emb)

oldhelp = None

def setup(bot):
    oldhelp = bot.remove_command("help")
    bot.add_cog(Help(bot))

def teardown(bot):
    bot.add_command(oldhelp)

pock/coggable/role.py
import discord
from discord.ext import commands
import random, asyncio

class StatusChanger(commands.Cog):

    def __init__(self, bot):
        self.bot = bot
        self.a = discord.Activity
        self.cancel = 0

    @commands.command()
    @commands.is_owner()
    async def begin(self, ctx):
        while self.cancel == 0:
            types = [discord.Game(name="with my uptime"),
                     discord.Activity(type=discord.ActivityType.listening, name=f"{len(self.bot.users)} cries of help"),
                     discord.Activity(type=discord.ActivityType.watching, name=f"{len(self.bot.guilds)} talk")]
            await self.bot.change_presence(random.choice(types))
            print("changed presence")
            await asyncio.sleep(random.choice([15, 30, 60]))


    @commands.command(name='break')
    @commands.is_owner()
    async def _break(self, ctx):
        """break the presence change and reset it

        blocker resets after 2 mins - begin must be run again."""
        self.cancel += 1
        activity = discord.Game(name='with {} servers'.format(len(self.bot.guilds)))
        await self.bot.change_presence(activity)
        m = await ctx.send("Successfully broke the chain.")
        await asyncio.sleep(120)
        await m.edit("Successfully broke the chain.\n\nReset - ready for use again.")


def setup(bot):
    bot.add_cog(StatusChanger(bot))

pock/coggable/listeners.py
import discord, datetime, traceback
from discord.ext import commands
from random import randint

class Ihaveears(commands.Cog):

    def __init__(self, bot):
        self.bot = bot
        self.dtf = "%I:%M %p @ %d/%m/%Y %Z"
        self.count = 0

    @commands.Cog.listener()
    async def on_guild_add(self, guild):
        to = self.bot.get_channel(543496620743065620)
        bots = 0
        members = 0
        e = discord.Embed(title="I joined a server!", description=datetime.datetime.utcnow().strftime(self.dtf),
                          timestamp=datetime.datetime.utcnow(), color=guild.owner.color)
        m = await to.send(embed=e)
        t = discord.Embed(title="I joined a server!", description=datetime.datetime.utcnow().strftime(self.dtf),
                          timestamp=datetime.datetime.utcnow(), color=guild.owner.color)
        for m in guild.members:
            if m.bot:
                bots += 1
            else:
                members += 1
        total = bots + members
        t.add_field(name="Member Count:", value="Bots: {}\nHumans: {}\nTotal: {}".format(bots, members, total))
        t.add_field(name="Owner:", value=guild.owner.name)
        await m.edit(embed=e)

    @commands.Cog.listener()
    async def on_command_error(self, ctx, error):
        if isinstance(error, discord.ext.commands.errors.CommandNotFound):
            return
        elif isinstance(error, commands.MissingPermissions):
            return await ctx.send(f"Error: missing permissions\n`{error}`")
        elif isinstance(error, commands.CommandOnCooldown):
            return await ctx.send(f"Command is on a cooldown! try again in `{round(error.retry_after, 2)}`s.")
        else:
            with open("log.txt", 'a+') as log:
                log.write(f"[ERROR] | {ctx.author} ran {ctx.command.name} in {ctx.guild.id} at "
                        f"{datetime.datetime.utcnow().strftime('%d/%m/%Y %I:%M%p%z')}\n")
                log.write(traceback.format_exc())
                log.close()
            await ctx.send(ctx.author.mention + " an unexpected error has occurred and has been logged. please report"
                                                " the below error to my developer:")
            await ctx.send(f"```py\n{error}\n```")
            if ctx.command.is_on_cooldown:
                cmd = ctx.command
                cmd.reset_cooldown(ctx)
            else:
                pass
            await ctx.send("Your error num: {}".format(randint(len(ctx.author.name), 999)))
            print("__________________________________________")
            raise error

    @commands.Cog.listener()
    async def on_command(self, ctx):
        with open("log.txt", 'a+') as l:
            l.write(f"[INFO] | {ctx.author} ran {ctx.command.name} in {ctx.guild.id} at "
                    f"{datetime.datetime.utcnow().strftime('%d/%m/%Y %I:%M%p%z')}\n")
            l.close()
        self.count += 1

    @commands.Cog.listener()
    async def on_error(self, error):
        raise error


def totalcommands(self, bot):
    bot.totalcommands = self.count

def setup(bot):
    bot.add_cog(Ihaveears(bot))


pock/coggable/maincog.py
import time
import datetime
import psutil
import os
from .listeners import totalcommands

from discord.ext import commands
import traceback
import discord
import textwrap
from contextlib import redirect_stdout
import io


class Smthn(commands.Cog):

    def __init__(self, bot):
        self.bot = bot
        # stats
        self.commands = totalcommands
        # else
        self.process = psutil.Process()
        self.toload = bot.cogss
        self._last_result = None
        self.dtf = "%I:%M %p @ %d/%m/%Y %Z"
    blankeval = '```\nprint("dude you need to tell me something to eval lul")\n```'

    @staticmethod
    def cleanup_code(self, content):
        """Automatically removes code blocks from the code."""
        # remove ```py\n```
        if content.startswith('```py' or '```') and content.endswith('```'):
            return '\n'.join(content.split('\n')[1:-1])
        return content.strip('` \n')

    @commands.command(aliases=['ui', 'u', 'user'])
    @commands.bot_has_permissions(embed_links=True)
    async def userinfo(self, ctx, user: discord.Member = None):
        """| Get @user/userID's info. kinda works cross-server."""
        dtformat = "%I:%M %p @ %d/%m/%Y %Z"
        if user is None:
            user = ctx.author
        u = user
        tmg = await ctx.send("Gathering info...", delete_after=5)
        msgs = "Idfk do `cmd + f` and `from: {}`".format(user)
        if u in ctx.guild.members:
            e = discord.Embed(title="{}'s info".format(u.name),
                              timestamp=datetime.datetime.utcnow(), color=ctx.author.color)
            e.add_field(name="Top Role:", value=u.top_role.name)
            e.add_field(name="Total Roles:", value=len(u.roles), inline=True)
            e.add_field(name="Created at:", value=u.created_at.strftime(dtformat), inline=False)
            e.add_field(name="Joined at:", value=u.joined_at.strftime(dtformat), inline=True)
            e.add_field(name="Total messages:", value=msgs)
            e.set_thumbnail(url=u.avatar_url)

            await ctx.send(embed=e)
        else:
            u = await self.bot.get_user_info(user.id)
            e = discord.Embed(title="{}'s info".format(u.name),
                              timestamp=datetime.datetime.utcnow(), color=ctx.author.color)
            e.add_field(name="Created at:", value=u.created_at.strftime(dtformat), inline=False)
            await ctx.send(embed=e)

    @commands.command()
    async def invite(self,ctx, dm: str = None):
        """Invite me"""
        if dm is None:
            mode = ctx
        elif dm == 'dm':
            mode = ctx.author
        elif dm == 'here':
            mode = ctx
        else:
            mode = ctx
        discgg = 'https://discord.gg/bwmVbdc'
        invlink = 'https://discordapp.com/oauth2/authorize?client_id=553622350952923146&permissions=0&scope=bot'
        e = discord.Embed(title=None, description="[Add the bot]({}) or, [join the server!]({})".format(invlink, discgg))
        try:
            await mode.send(embed=e)
        except discord.Forbidden:
            await mode.send(embed=e)

    @commands.command(name='perms', aliases=['perms_for', 'permissions'])
    @commands.guild_only()
    async def check_permissions(self, ctx, *, member: discord.Member=None):
        """A simple command which checks a members Guild Permissions.
        If member is not provided, the author will be checked."""

        if not member:
            member = ctx.author

        # Here we check if the value of each permission is True.
        perms = '\n'.join(perm for perm, value in member.guild_permissions if value)

        # And to make it look nice, we wrap it in an Embed.
        embed = discord.Embed(title='Permissions for:', description=ctx.guild.name, colour=member.colour)
        embed.set_author(icon_url=member.avatar_url, name=str(member))

        # \uFEFF is a Zero-Width Space, which basically allows us to have an empty field name.
        embed.add_field(name=None, value=perms)
        await ctx.send(content=None, embed=embed)

    @commands.group()
    @commands.is_owner()
    async def console(self, ctx):
        pass
    @console.command()
    @commands.is_owner()
    async def clear(self, ctx):
        """Clears console"""
        nl = '\n'
        print(nl*999)
        print(f"Console cleared by {ctx.author}.\n__________________________________________")
    @console.command(name='log')
    @commands.is_owner()
    async def ulog(self, ctx):
        """Upload the log file."""
        await ctx.trigger_typing()
        await ctx.send(file=discord.File('log.txt'))
    @console.command()
    @commands.is_owner()
    async def status(self, ctx):
        """Reset the status"""
        await self.bot.change_presence(activity=discord.Game(name='with {} servers'.format(len(self.bot.guilds))))
        await ctx.send("Success.")
    @console.command()
    @commands.is_owner()
    async def debug(self, ctx):
        """get usage stats"""
        embed=discord.Embed(title="sys stats", color=discord.Color.blurple())
        memory_usage = self.process.memory_full_info().uss / 1024**2
        corenum = 0
        forefmt = ""
        cpu_usage = psutil.cpu_percent(interval=None, percpu=True)
        for corepercent in cpu_usage:
            corenum += 1
            forefmt += f"Core {corenum}: {corepercent}%\n"
        embed.add_field(name='Process', value=f'{memory_usage:.2f} MiB\nCPU: {forefmt}')
        embed.add_field(name="Total mem", value=psutil.virtual_memory().percent)
        await ctx.send(embed=embed)
    @console.group()
    @commands.is_owner()
    async def cmd(self, ctx):
        """IDFK"""
    @cmd.command()
    @commands.is_owner()
    async def remcd(self, ctx, commandname: str):
        """Remove the cooldown on a command."""
        cmd = self.bot.get_command(commandname)
        if not cmd:
            return ctx.send("i can't insta-cool a command that doesn't exist!")
        else:
            try:
                cmd.reset_cooldown(ctx)
                return ctx.send("Reset {}'s cooldown.".format(commandname))
            except Exception as e:
                return ctx.send(e)
    @console.command()
    @commands.is_owner()
    async def eval(self, ctx, body: str = f'{blankeval}'):
        """Eval code"""
        env = {
            'bot': self.bot,
            'ctx': ctx,
            'channel': ctx.channel,
            'author': ctx.author,
            'guild': ctx.guild,
            'message': ctx.message,
            '_': self._last_result
        }
        env.update(globals())
        body = self.cleanup_code(self, content=body)
        stdout = io.StringIO()
        to_compile = f'async def func():\n{textwrap.indent(body, "  ")}'
        try:
            exec(to_compile, env)
        except Exception as e:
            return await ctx.send(f'```py\n{e.__class__.__name__}: {e}\n```')
        func = env['func']
        try:
            with redirect_stdout(stdout):
                ret = await func()
        except Exception as e:
            value = stdout.getvalue()
            await ctx.send(f'```py\n{value}{traceback.format_exc()}\n```')
        else:
            value = stdout.getvalue()
            try:
                await ctx.message.add_reaction('\u2705')
            except:
                pass
            if ret is None:
                if value:
                    await ctx.send(f'```py\n{value}\n```')
            else:
                self._last_result = ret
                await ctx.send(f'```py\n{value}{ret}\n```')

    @console.command()
    @commands.is_owner()
    async def servers(self, ctx):
        """List servers"""
        servers = "```\n"
        for c, s in enumerate(self.bot.guilds):
            c = c + 1
            servers += f"{c} {s.name} | {s.id}\n"
        servers += f"Total: {len(self.bot.guilds)}\n"
        servers += "```"
        print(servers)
        await ctx.send(servers)

    @commands.command()
    @commands.bot_has_permissions(manage_messages=True)
    @commands.has_permissions(manage_messages=True)
    @commands.cooldown(1, 5, commands.BucketType.guild)
    async def purge(self, ctx, amount: int = 100):
        """Get them messages outta here!"""
        try:
            await ctx.message.delete()
            d = await ctx.channel.purge(limit=amount, bulk=True)
            await ctx.send("{}, i successfully purged {} messages.".format(ctx.author.mention, len(d)))
        except:
            try:
                d = await ctx.channel.purge(limit=amount, bulk=False)
                await ctx.send("{}, i successfully purged {} messages.".format(ctx.author.mention, len(d)))
            except:
                return ctx.send("Error: i need `manage_messages`. not sure how you ran the cmd without giving me it, "
                                "but ok.")

    @commands.command()
    @commands.bot_has_permissions(read_message_history=True)
    @commands.bot_has_permissions(attach_files=True)
    @commands.has_permissions(manage_messages=True)
    async def chatlog(self, ctx, limit: int = 1000):
        """Get [limit] amount of messages and output in a nice text file.

        the limit is capped at 999k, but the larger said limit is, the longer it will take to make and upload."""
        tm = await ctx.send("logging...")
        if limit >= 999999:
            limit = 999999
        async with ctx.channel.typing():
            amount = 0
            start = time.time()
            with open(f"{ctx.guild.name}-{ctx.channel.name}.txt", "a+") as l:
                async for m in ctx.channel.history(limit=limit, before=ctx.message.created_at, reverse=True):
                    amount += 1
                    l.write(f"{m.author} at {m.created_at.strftime('%d/%m/%Y %I:%M%p %z')}:\n{m.content}\n")
            l.close()
            end = time.time()
        await tm.edit(content="Logged {} messages in {} seconds\nFile:".format(
            amount, round(end - start, 2)))
        await ctx.send(file=discord.File(f"{ctx.guild.name}-{ctx.channel.name}.txt"))
        os.remove(f"{ctx.guild.name}-{ctx.channel.name}.txt")

    @commands.command(aliases=['announce', 'sendembed'])
    @commands.has_permissions(embed_links=True)
    @commands.bot_has_permissions(embed_links=True)
    async def embed(self, ctx):
        """Embed a message!
        DO NOT PASS ANY ARGS WHEN EXECUTING"""
        a = ctx.author.mention
        m = await ctx.send(f"{a}, this command is WIP. sorry")

    @commands.command()
    @commands.bot_has_permissions(embed_links=True)
    async def stats(self, ctx, brief: bool = True):
        """Get bot stats.

        say `//stats False` to get lengthy, more in-depth info"""
        o = await self.bot.get_user_info(421698654189912064)
        url = 'http://discord.gg/mNV6azN'
        if brief:
            e = discord.Embed(title="My info!", color=self.bot.user.color, timestamp=datetime.datetime.utcnow())
            e.url = url
            e.add_field(name="Users:", value=len(self.bot.users))
            cc = 0
            for s in self.bot.guilds:
                cc += len(s.channels)
            e.add_field(name="Channels:", value=cc)
            e.add_field(name="Guilds:", value=len(self.bot.guilds))
            e.add_field(name="Total executed commands:", value="Null")
            e.add_field(name="Owner:", value=o)
            await ctx.send(embed=e)
        else:
            loaded = 0
            cogs = 4
            loadedcogs = 0
            musage = psutil.virtual_memory().percent
            corenum = 0
            forefmt = ""
            cpu_usage = psutil.cpu_percent(interval=None, percpu=True)
            for corepercent in cpu_usage:
                corenum += 1
                forefmt += f"Core {corenum}: {corepercent}%\n"
            e = discord.Embed(title="My info!", color=self.bot.user.color, timestamp=datetime.datetime.utcnow())
            e.url = url
            e.add_field(name="Users:", value=len(self.bot.users))
            cc = 0
            files = 0
            for s in self.bot.guilds:
                cc += len(s.channels)
            for d in os.listdir():
                files += 1
            e.add_field(name="Channels:", value=cc)
            e.add_field(name="Guilds:", value=len(self.bot.guilds))
            e.add_field(name="Total executed commands:", value="Null")
            e.add_field(name="Owner:", value=o)
            e.add_field(name="Memory usage:", value=f"{musage}%")
            e.add_field(name="CPU usage:", value=forefmt)
            e.add_field(name="Total files:", value=files)
            for c in self.toload:
                try:
                    self.bot.load_extention(c)
                    loadedcogs += 1
                except:
                    loadedcogs += 1
            e.add_field(name="Cogs:", value=f"{loadedcogs}/{cogs}")
            await ctx.send(embed=e)

    @commands.command()
    @commands.bot_has_permissions(create_instant_invite=True)
    async def geninv(self, ctx, channel: discord.TextChannel = None):
        """Generate an invite. mention a channel to remote create """
        if not channel:
            channel = ctx.channel
        if ctx.author.id == 421698654189912064:
            t = await channel.create_invite(reason="Invoked by owner.")
            await ctx.send(t.url)
        else:
            t = channel.create_invite(max_age=604800, reason="Invoked by {}".format(ctx.author.name))
            await ctx.send(t.url)

    @commands.command()
    @commands.bot_has_permissions(embed_links = True)
    async def serverinfo(self, ctx, id: int = None):
        """Get server infos. supply an id for another server

        note for id the bot must be *in* the server."""
        async with ctx.typing():
            invites = ""
            bots = 0
            humans = 0
            if id is None:
                server = ctx.guild
            else:
                server = self.bot.get_server(id)
            try:
                invs = await server.invites()
                for s in invs:
                    if s.max_age and s.max_uses == 0:
                        invites += s.url
                        break
            except Exception as e:
                invites += "https://www.daimto.com/wp-content/uploads/2014/08/errorstop.png"
                print(e)
            e = discord.Embed(title=f"{server.name}'s info!", timestamp=datetime.datetime.utcnow(), color=server.owner.color,
                              description=f"Owner: {server.owner} ({server.owner.mention})")
            e.url=invites
            for u in server.members:
                if u.bot:
                    bots += 1
                else:
                    humans += 1
            total = bots + humans
            e.add_field(name="Members:", value=f"`{bots}` bots\n{humans} humans\n{total} total")
            e.add_field(name="Roles:", value=str(len(server.roles)))
            # the str(len()) is so pep8 doesnt yell at me
            e.add_field(name="Channels:", value=str(len(server.channels)))
            e.add_field(name="Emoji(s):", value=str(len(server.emojis)))
            e.add_field(name="Created at:", value=server.created_at.strftime(self.dtf))
            e.set_thumbnail(url=server.icon_url)
        await ctx.send(embed=e)
def setup(bot):
    bot.add_cog(Smthn(bot))
